### PR TITLE
fix(crates): do not process relative dependencies twice

### PR DESCRIPTION
**Pull request by Maddiaa0**
# Description

Currently in the aztec3 repo we are unable to import helper dependencies that rely on the `aztec3-noir` standard library that we have created. 

The module structure is something like this:
```
- root
- library
- helper
```

in our case both helper depends on library and helper, where helper _also_ depends on library. This leads to noir compiling the library files, then recompiling them when processing the helper. 

This leads to instances like below, where the same type has two different idents that mismatch.
image

This pr fixes this issue by recognising when a dependant local crate has already been compiled, and not compiling it again

## Problem\*

Resolves 

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/fm/src/lib.rs [hunk 1]
```diff
@@ -38,12 +38,21 @@ pub struct FileManager {
 impl FileManager {
     // XXX: Maybe use a AsRef<Path> here, for API ergonomics
     pub fn add_file(&mut self, path_to_file: &Path, file_type: FileType) -> Option<FileId> {
-        let source = file_reader::read_file_to_string(path_to_file).ok()?;
+        // Handle both relative file paths and std/lib virtual paths.
+        let base = Path::new(".").canonicalize().unwrap(); // Should never fail (TODO: test wasm)
```
Comment (ludamad) [inline crates/fm/src/lib.rs:42]: I prefer 
        .expect("Base path canonicalization failed"); to unwrap

File: crates/noirc_frontend/src/graph/mod.rs [hunk 2]
```diff
@@ -83,7 +83,13 @@ impl CrateGraph {
...
         let mut roots_with_file_id =
             self.arena.iter().filter(|(_, crate_data)| crate_data.root_file_id == file_id);
-        assert!(roots_with_file_id.next().is_none(), "you cannot add the same file id twice");
+
+        let next_file_id = roots_with_file_id.next();
+        assert!(next_file_id.is_none(), "you cannot add the same file id twice");
+
+        if let Some(file_id) = next_file_id {
+            return *file_id.0;
+        }
```
Comment (jfecher) [inline crates/noirc_frontend/src/graph/mod.rs:92]: Looks like this doesn't do anything. We assert the Option is None first and if not, we panic. Then we check if it is Some afterward, although we already would have panicked by this point.

File: crates/fm/src/lib.rs [hunk 3]
```diff
@@ -38,12 +38,21 @@ pub struct FileManager {
 impl FileManager {
     // XXX: Maybe use a AsRef<Path> here, for API ergonomics
     pub fn add_file(&mut self, path_to_file: &Path, file_type: FileType) -> Option<FileId> {
-        let source = file_reader::read_file_to_string(path_to_file).ok()?;
+        // Handle both relative file paths and std/lib virtual paths.
+        let base = Path::new(".").canonicalize().expect("Base path canonicalize failed"); // Should never fail
```
Comment (jfecher) [inline crates/fm/src/lib.rs:42]:
```suggestion
        let base = Path::new(".").canonicalize().expect("Base path canonicalize failed");
```
We can remove the comment, we should assume any call to `expect` shouldn't fail. If we expect it to fail, it should be a user-facing error message with source location, etc.

File: crates/fm/src/lib.rs [hunk 4]
```diff
@@ -38,12 +38,21 @@ pub struct FileManager {
...
+        let res = path_to_file.canonicalize().unwrap_or_else(|_| path_to_file.to_path_buf());
+        let resolved_path = res.strip_prefix(base).unwrap_or(&res);
```
Comment (jfecher) [inline crates/fm/src/lib.rs:44]: Retrieving the current directory beforehand with `let base = Path::new(".").canonicalize()`, then calling `canonicalize` on `path_to_file` as well just to strip the base prefix afterward seems a bit roundable.

Can we use absolute paths everywhere instead? E.g. let's use `path_to_file.canonicalize()` without stripping any prefixes.

File: crates/noirc_frontend/src/graph/mod.rs (hunk 2 above)
Comment (Maddiaa0) [reply crates/noirc_frontend/src/graph/mod.rs]: i've removed this check

**Comment by jfecher** [comment]
Looks like `graph::tests::it_works2` is failing

File: crates/noirc_frontend/src/graph/mod.rs [hunk 5]
```diff
@@ -83,7 +83,11 @@ impl CrateGraph {
     pub fn add_crate_root(&mut self, crate_type: CrateType, file_id: FileId) -> CrateId {
         let mut roots_with_file_id =
             self.arena.iter().filter(|(_, crate_data)| crate_data.root_file_id == file_id);
-        assert!(roots_with_file_id.next().is_none(), "you cannot add the same file id twice");
```
Comment (jfecher) [inline crates/noirc_frontend/src/graph/mod.rs:86]: So were we hitting this assert?

Comment (Maddiaa0) [reply crates/noirc_frontend/src/graph/mod.rs:86]: yep, it must have hit it before as it was adding the same path twice

**Review by jfecher — APPROVED** [review]
👍

**Comment by phated** [comment]
I believe this broke wasm because (last time I checked) the canonicalize function is not supported by wasi. I had replaced this in circom with an alternative impl.

**Pull request merged**